examples/Custom_Hooks/probability_of_improvement.py

```diff
@@ -0,0 +1,149 @@
+## Plot the Probability of Improvement
+
+# This example demonstrates how to use the
+# {func}`register_hooks <baybe.utils.basic.register_hooks>` utility,
+# to calculate and analyse the probability of improvement:
+# * We define a hook that is compatible with the general
+#   {meth}`RecommenderProtocol.recommend <baybe.recommenders.base.RecommenderProtocol.recommend>`
+#   interface,
+# * attach it to a recommender,
+# * and watch it take action.
+
+
+### Imports
+
+
+from types import MethodType
+
+import matplotlib.pyplot as plt
+import pandas as pd
+import seaborn as sns
+
+from baybe.acquisition import ProbabilityOfImprovement
+from baybe.campaign import Campaign
+from baybe.objectives.base import Objective
+from baybe.objectives.single import SingleTargetObjective
+from baybe.parameters import CategoricalParameter, NumericalDiscreteParameter
+from baybe.recommenders import (
+    BotorchRecommender,
+    RandomRecommender,
+    TwoPhaseMetaRecommender,
+)
+from baybe.searchspace import SearchSpace, SearchSpaceType
+from baybe.surrogates import GaussianProcessSurrogate
+from baybe.targets import NumericalTarget
+from baybe.utils.basic import register_hooks
+from baybe.utils.dataframe import add_fake_results, to_tensor
+
+### Setup
+
+# We start by initialize the global poi_list where we will store the probability of improvement from each recommend iteration
+
+poi_list = []
+
+
+# Then, we define the hook that calculates the probability of improvement based on the given recommender, search space and measurements:
+
+
+def check_probability_of_improvement(
+    self: BotorchRecommender,
+    searchspace: SearchSpace,
+    objective: Objective | None = None,
+    measurements: pd.DataFrame | None = None,
+):
+    """Calculate and store the probability of improvement in poi_list."""
+    if searchspace.type == SearchSpaceType.DISCRETE:
+        train_x = searchspace.transform(measurements)
+        train_y = objective.transform(measurements)
+        ac_f = ProbabilityOfImprovement()
+        botorch_acf = ac_f.to_botorch(
+            self.surrogate_model, searchspace, train_x, train_y
+        )
+        comp_rep_tensor = to_tensor(searchspace.discrete.comp_rep).unsqueeze(1)
+        poi = botorch_acf(comp_rep_tensor)
+        poi_list.append(poi)
+    else:
+        raise TypeError(
+            f"{searchspace.type} search spaces are not supported yet. "
+            f"Currently only DISCRETE search spaces are accepted."
+        )
+
+
+# Additionally, we define a function that plots the poi_list after all recommend  iterations:
+
+
+def plot_poi(poi_list: list):
```

Comment:
so far it is not exactly what I had in mind but let us discuss the plot first in person